graphql_compiler/compiler/emit_cypher.py

```diff
@@ -0,0 +1,139 @@
+    no_linked_location = cypher_step.linked_location is None
+    step_is_query_root = isinstance(cypher_step.step_block, QueryRoot)
+
+    if no_linked_location ^ step_is_query_root:
+        raise AssertionError(u'Received an illegal CypherStep. Not having a linked location is '
+                             u'allowed if and only if the step is with a QueryRoot object. {}'
+                             .format(cypher_step))
+
+    has_where_block = cypher_step.where_block is not None
+    is_optional_step = (
+        isinstance(cypher_step.step_block, Traverse) and
+        cypher_step.step_block.optional
+    )
+    if has_where_block and is_optional_step:
+        raise AssertionError(u'Received an illegal CypherStep containing an optional step together '
+                             u'with a "where" Filter block. This is a bug in the lowering code, as '
+                             u'it should have moved the filtering to the global operations '
+                             u'section. {}'.format(cypher_step))
+
+    step_location = cypher_step.as_block.location
+    step_location_name, _ = step_location.get_location_name()
+
+    template_data = {
+        'step_location': step_location_name,
+        'step_vertex_type': u':'.join(sorted(cypher_step.step_types)),
+        'quantifier': u'',
+        'left_edge_mark': u'',
+        'right_edge_mark': u'',
+    }
+    step_vertex_pattern = u'(%(step_location)s:%(step_vertex_type)s)'
+
+    if cypher_step.linked_location is None:
+        pattern = u'MATCH ' + step_vertex_pattern
+    else:
+        pattern = (
+            u'MATCH (%(linked_location)s)'
+            u'%(left_edge_mark)s-[:%(edge_type)s%(quantifier)s]-%(right_edge_mark)s' +
+            step_vertex_pattern
+        )
+        linked_location_name, _ = cypher_step.linked_location.get_location_name()
+        template_data['linked_location'] = linked_location_name
+
+    if has_where_block:
+        pattern += u'\n  WHERE %(predicate)s'
+        template_data['predicate'] = cypher_step.where_block.predicate.to_cypher()
+
+    if is_optional_step:
+        pattern = u'OPTIONAL ' + pattern
+
+    if isinstance(cypher_step.step_block, (Traverse, Recurse, Fold)):
+        if isinstance(cypher_step.step_block, Fold):
+            direction, edge_name = cypher_step.step_block.fold_scope_location.fold_path[0]
+        else:
+            direction = cypher_step.step_block.direction
+            edge_name = cypher_step.step_block.edge_name
+
+        template_data['edge_type'] = edge_name
+
+        direction_lookup = {
+            'in': ('left_edge_mark', u'<'),
+            'out': ('right_edge_mark', u'>'),
+        }
+        key, value = direction_lookup[direction]
+        template_data[key] = value
+
+    if isinstance(cypher_step.step_block, Recurse):
+        template_data['quantifier'] = u'*0..%d' % cypher_step.step_block.depth
+
+    # Try to obey the Cypher style guidebook at least a little bit.
```

Comment:
Way too many words to communicate conformity to the guidebook.
"Comply with Cypher style guidebook on whitespace a bit.", perhaps?

Reply:
sounds good to me